### PR TITLE
Provides almost complete abstract implementations.

### PR DESCRIPTION
**Pull request by darora**
Make it easier to selectively provide features in ls implementations

### CODE REVIEW

File: language-server-commons/src/main/java/com/palantir/ls/services/AbstractWorkspaceService.java [hunk 1]
```diff
@@ -43,4 +44,8 @@ public final void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
         getState().publishDiagnostics(getState().getCompilerWrapper().compile());
     }
 
+    @Override
+    public void didChangeConfiguraton(DidChangeConfigurationParams didChangeConfigurationParams) {
```
Comment (ags799) [inline language-server-commons/src/main/java/com/palantir/ls/services/AbstractWorkspaceService.java:48]: `didChangeConfiguration`

**Comment by darora** [comment]
Yup, still cleaning up this PR for baseline as well, will mark it ready to review once I'm done.

File: language-server-commons/src/main/java/com/palantir/ls/services/AbstractWorkspaceService.java (hunk 1 above)
Comment (darora) [reply language-server-commons/src/main/java/com/palantir/ls/services/AbstractWorkspaceService.java:48]: hmm?

Comment (ags799) [reply language-server-commons/src/main/java/com/palantir/ls/services/AbstractWorkspaceService.java:48]: spelling, `didChangeConfiguraton` -> `didChangeConfiguration`

Comment (darora) [reply language-server-commons/src/main/java/com/palantir/ls/services/AbstractWorkspaceService.java:48]: Must have looked at it six times before you posted that, and didn't see it

Comment (jiahuijiang) [reply language-server-commons/src/main/java/com/palantir/ls/services/AbstractWorkspaceService.java:48]: ahh lol this is in ls-api, will submit a PR with them

**Comment by jiahuijiang** [comment]
@darora @bavardage @ags799 @gracew
Did some more refactoring and moved classes around so that SQL language server won't need to have all the copied code.
Fixed all the findbugs warnings, apparently in the new baseline the instructions on when to use final classes have changed a lot.

File: .baseline/findbugs/excludeFilter.xml [hunk 2]
```diff
@@ -1,6 +1,10 @@
 <!-- See http://findbugs.sourceforge.net/manual/filter.html for syntax -->
 <FindBugsFilter>
 
+    <Match>
+        <Source name="~.*\.scala" />
```
Comment (gracew) [inline .baseline/findbugs/excludeFilter.xml:5]: why is this here. 🚫  scala!!

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServer.java [hunk 3]
```diff
@@ -47,7 +47,7 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public final class GroovyLanguageServer implements LanguageServer {
+public class GroovyLanguageServer implements LanguageServer {
```
Comment (gracew) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServer.java:50]: this should remain final?

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/Indexer.java [hunk 4]
```diff
@@ -26,7 +26,7 @@
 import java.util.Map;
 import java.util.Set;
 
-public final class Indexer {
+public class Indexer {
```
Comment (gracew) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/Indexer.java:29]: I'm not sure why any of the groovy classes need to be non-final

File: language-server-commons/src/main/java/com/palantir/ls/DefaultLanguageServerState.java [hunk 5]
```diff
@@ -2,7 +2,7 @@
  * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
+ * you may not use this file except in compliance ctes the License.
```
Comment (gracew) [inline language-server-commons/src/main/java/com/palantir/ls/DefaultLanguageServerState.java:5]: whoops

File: language-server-commons/src/main/java/com/palantir/ls/api/TreeParser.java [hunk 6]
```diff
@@ -67,4 +75,15 @@
      */
     Set<SymbolInformation> getFilteredSymbols(String query);
 
+    default Pattern getQueryPattern(String query) {
+        String escaped = Pattern.quote(query);
+        String newQuery = escaped.replaceAll("\\*", "\\\\E.*\\\\Q").replaceAll("\\?", "\\\\E.\\\\Q");
```
Comment (gracew) [inline language-server-commons/src/main/java/com/palantir/ls/api/TreeParser.java:80]: what is this? What is E/Q?

File: language-server-commons/src/test/java/com/palantir/ls/services/DefaultTextDocumentServiceTest.java [hunk 7]
```diff
@@ -77,8 +78,7 @@
 import org.mockito.Mock;
 import org.mockito.MockitoAnnotations;
 
-// TODO (darora): #131 split this up and test things independently in the language-server-commons as well
-public final class DefaultTextDocumentServiceTest {
+public class DefaultTextDocumentServiceTest {
```
Comment (gracew) [inline language-server-commons/src/test/java/com/palantir/ls/services/DefaultTextDocumentServiceTest.java:81]: lol I'm guessing now that you did a find and replace all instances of "public final class"...

**Review by gracew — COMMENTED** [review]
I'll submit a flup that addresses my (very belated) comments

File: .baseline/findbugs/excludeFilter.xml (hunk 2 above)
Comment (gracew) [reply .baseline/findbugs/excludeFilter.xml:5]: oh sorry, didn't realize this is now part of baseline

**Comment by darora** [comment]
@gracew all the comments you have are baseline-caused. Except for my glorious `ctes`

**Comment by gracew** [comment]
meh ok, I see that all the omissions of final (other than test classes, which I care less about) are due to https://github.com/palantir/gradle-baseline/pull/61

**Comment by gracew** [comment]
yep 👍

File: .baseline/eclipse/static/dotfile.settings/org.eclipse.jdt.ui.prefs [hunk 8]
```diff
@@ -11,7 +11,7 @@ org.eclipse.jdt.ui.keywordthis=false
...
...tchblock_context" deleted\="false" description\="Code in new catch blocks" enabled\="true" id\="org.eclipse.jdt.ui.text.codetemplates.catchblock" name\="catchblock">throw new RuntimeException(${exception_var});</template><template autoinsert\="false" context\="methodbody_context" deleted\="false" description\="Code in created method stubs" enabled\="true" id\="org.eclipse.jdt.ui.text.codetemplates.methodbody" name\="methodbody">${body_statement}</template><template autoinsert\="false" context\="constructorbody_context" deleted\="false" description\="Code in created constructor stubs" enabled\="true" id\="org.eclipse.jdt.ui.text.codetemplates.constructorbody" name\="constructorbody">${body_statement}</template><template autoinsert\="true" context\="getterbody_context" deleted\="false" description\="Code in created getters" enabled\="true" id\="org.eclipse.jdt.ui.text.codetemplates.getterbody" name\="getterbody">return ${field};</template><template autoinsert\="true" context\="setterbody_context" deleted\="false" description\="Code in created setters" enabled\="true" id\="org.eclipse.jdt.ui.text.codetemplates.setterbody" name\="setterbody">${field} \= ${param};</template></templates>
```
Comment (gracew) [inline .baseline/eclipse/static/dotfile.settings/org.eclipse.jdt.ui.prefs:14]: the baseline upgrade overwrote the apache license here... see #53 for the previous PR tweaking this
